### ISSUE TITLE
Add ability to cleanup Configuration related disposables 

### ISSUE THREAD

**Issue by Eilon** [dotnet/extensions#786]
_From @JanEggers on Friday, 28 September 2018 06:43:27_

Im having issues in a Unittests that complains that not all threads are ended when the unittest is completed.

Turns out this is due to using this:

```
 var configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json")
                .AddJsonFile(settingsPath, optional: true, reloadOnChange: true)
                .Build();
```

the reloadable json file generates Filewatchers that hang in a sleep if the test is ended.

the core problem is that there PhsicalFileProvider is disposable but currently I have to do something like this to clean it up:

```
foreach (var provider in config.Providers.OfType<JsonConfigurationProvider>())
            {
                if (provider.Source.FileProvider is PhysicalFileProvider pfp)
                {
                    pfp.Dispose();
                }
            }
```

I see two options here to improve the situation:

1. make IConfigurationRoot and all related classes disposable so I can just call config.Dispose();

2. resolve providers from DI so DI takes care of disposing the resources that require to do so. 

maybe related to #3110 

_Copied from original issue: aspnet/Home#3564_

**Comment by davidfowl** [comment]
Yes we need to support disposable configuration providers which is viral so the root will need to support it as well.

**Comment by philip-reed** [comment]
I have just ran into I think the same issue when trying to implement a Custom Configuration provider:
[https://stackoverflow.com/questions/54072129/stop-sqldependancy-in-custom-aspnet-core-configuration-provider](https://stackoverflow.com/questions/54072129/stop-sqldependancy-in-custom-aspnet-core-configuration-provider) 

Whatever is implemented to dispose of the FileWatcher/FileProvider bits, it would be useful if something was exposed to also allow custom providers to be disposed also.

**Comment by poke** [comment]
This issue can be closed. It’s fixed with the PRs #928, #1361, and aspnet/AspNetCore#9149, and should already work in preview 5.

**Comment by anurse** [comment]
Closing based on the previous comment as part of backlog clean-up.
